firecracker-control/local.go

```diff
@@ -77,20 +80,38 @@ func newLocal(ic *plugin.InitContext) (*local, error) {
 	return &local{
 		vm:          make(map[string]*instance),
 		rootPath:    ic.Root,
+		publisher:   ic.Events,
 		findVsockFn: findNextAvailableVsockCID,
 	}, nil
 }
 
 // CreateVM creates new Firecracker VM instance
-func (s *local) CreateVM(ctx context.Context, req *proto.CreateVMRequest) (*proto.CreateVMResponse, error) {
-	id, err := s.makeID()
+func (s *local) CreateVM(ctx context.Context, req *proto.CreateVMRequest) (*empty.Empty, error) {
+	var (
+		id     = req.GetVMID()
+		logger = log.G(ctx).WithField("vm_id", id)
+	)
+
+	if id == "" {
+		return nil, errors.New("invalid VM ID")
+	}
+
+	s.vmLock.Lock()
```

Comment:
So basically this method is non-concurrent. Is this due to finding the vsock CID?

Reply:
See Erik's comment: https://github.com/firecracker-microvm/firecracker-containerd/pull/163#discussion_r275515176

Reply:
Ah, if it is only racing on the map, I would much prefer the use of the `sync.Map` structure.

Reply:
That's not going to work. we need to check+add id atomically in order to prevent race conditions.

Reply:
Use the `LoadStore` and if it already exists, it'll return true and you can return an error. This has a smaller critical section

Reply:
This gives same results as with mutex. See my thoughts regarding `sync.Map` here: https://github.com/firecracker-microvm/firecracker-containerd/pull/156#discussion_r273284575

Reply:
But it is also smaller amounts of code and prevents a smaller critical section. I would be in favor of changing this since it is relatively small and easy. Right now, `CreateVM` is a sequential API.

Reply:
How is it going to be smaller?

Reply:
You don't need the first if statement for one. The second if statement will be a `LoadStore` check and then cast. So my guess is one line less :p. But more importantly smaller critical section.

Reply:
So you create a VM first and than call `LoadStore` at the end?